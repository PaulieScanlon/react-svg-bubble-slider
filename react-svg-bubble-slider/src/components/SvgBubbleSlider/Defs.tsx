import React, { FunctionComponent, memo } from 'react'

export const Defs: FunctionComponent = memo(() => (
  <defs>
    <filter id="goo" colorInterpolationFilters="sRGB">
      <feGaussianBlur in="SourceGraphic" stdDeviation="8" result="blur" />
      <feColorMatrix
        in="blur"
        mode="matrix"
        values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 31 -12"
        result="cm"
      />
    </filter>
  </defs>
))

Defs.displayName = 'Defs'
