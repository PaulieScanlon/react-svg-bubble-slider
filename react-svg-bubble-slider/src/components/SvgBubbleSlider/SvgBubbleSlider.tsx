import React, { Fragment, FunctionComponent, useState } from 'react'

import { SvgBubbleSliderProps } from './types'
import theme from '../../theme'

import { Timeline } from './Timeline'

export const SvgBubbleSlider: FunctionComponent<SvgBubbleSliderProps> = ({
  children,
  primaryColor = theme.colors.primary,
  secondaryColor = theme.colors.background,
  icons,
  showSpeechBubble,
}: SvgBubbleSliderProps) => {
  const [currentReaction, setCurrentReaction] = useState<any>('')

  return (
    <Fragment>
      <div
        className="react-svg-bubble-slider"
        style={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'center',
          position: 'relative',
          width: '100%',
        }}
      >
        <Timeline
          onAnimationComplete={(reaction) => setCurrentReaction(reaction)}
          primaryColor={primaryColor}
          secondaryColor={secondaryColor}
          icons={icons}
          showSpeechBubble={showSpeechBubble}
        />
      </div>
      <div className="svg-bubble-action">
        {children &&
          children({
            reaction: currentReaction,
          })}
      </div>
    </Fragment>
  )
}
