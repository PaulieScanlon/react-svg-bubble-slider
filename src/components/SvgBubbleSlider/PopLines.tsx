import React, { FunctionComponent, forwardRef, RefObject } from 'react'

interface PopLinesProps {
  /** The colors of the lines */
  color: string
  /** The size of the viewbox */
  viewboxWidth: number
  /** Ah ref!!! */
  ref: any
}

const SELF_WIDTH = 170

export const PopLines: FunctionComponent<PopLinesProps> = forwardRef<
  any,
  PopLinesProps
>(({ color, viewboxWidth }, ref) => (
  <g
    ref={ref as any}
    className="pop-Lines"
    fill="none"
    stroke={color}
    strokeWidth="4"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeMiterlimit="10"
    transform={`matrix(2,0,0,2,${viewboxWidth / 2 - SELF_WIDTH},0)`}
  >
    <line x1="107.923" y1="21.37" x2="116.935" y2="4"></line>
    <line x1="59.335" y1="24.909" x2="50.5" y2="8.057"></line>
    <line x1="21.789" y1="56.11" x2="4" y2="53.67"></line>
    <line x1="26.782" y1="98.86" x2="8.057" y2="108.34"></line>
    <line x1="65.885" y1="125.86" x2="56.43" y2="145.37"></line>
    <line x1="112.429" y1="121.342" x2="121.03" y2="142.564"></line>
    <line x1="147.535" y1="93.82" x2="168.155" y2="101.45"></line>
    <line x1="149.742" y1="49.01" x2="168.155" y2="42.56"></line>
  </g>
))
