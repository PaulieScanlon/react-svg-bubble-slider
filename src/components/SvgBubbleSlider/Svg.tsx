import React, {
  FunctionComponent,
  memo,
  RefObject,
  useEffect,
  useRef,
  useState,
  useMemo,
} from 'react'

import { gsap } from 'gsap'
import { Draggable } from 'gsap/Draggable'
import { InertiaPlugin } from 'gsap/InertiaPlugin'

gsap.registerPlugin(Draggable, InertiaPlugin)

import { iconPaths } from './iconPaths'

const ICON_SIZE = 32
const ICON_FILL = '#ffffff'
const MULTIPLIER = 4.8
const SPACER = 60

const DOT_SIZE = 10
const DOT_FILL = '#FF69B4'

const MIN_DRAG_X = -(iconPaths.length - 1) * SPACER
const VIEWBOX_WIDTH = SPACER * iconPaths.length

interface SvgProps {
  /** A percentage scale */
  scale: string
  /** Animation callback passes current reaction */
  onAnimationComplete: (reaction: string) => void
}

export const Svg: FunctionComponent<SvgProps> = memo(
  ({ scale, onAnimationComplete }: SvgProps) => {
    const svgRef = useRef(null)
    const dotContainerRef = useRef(null)
    const iconContainerRef = useRef(null)
    const iconRefs: RefObject<SVGPathElement>[] = []
    const dotRefs: RefObject<SVGCircleElement>[] = []

    const [hasMounted, sethasMounted] = useState(false)

    const [snapArray, setSnapArray] = useState([])

    const mtl = useMemo(() => gsap.timeline({ paused: true }), null)
    const [tls, setTls] = useState([])

    const handleDragSlider = () => {
      const posX = Number(gsap.getProperty(dotContainerRef.current, 'x'))

      gsap.to(mtl, {
        duration: 0.5,
        time: (posX / MIN_DRAG_X) * (mtl.duration() - 2) + 1,
        ease: 'elastic(2, 0.75)',
      })

      gsap.set(iconContainerRef.current, {
        x: posX,
      })
    }

    const handleDragStart = () => {
      // console.log('handleDragStart')
    }

    const handleThrowComplete = () => {
      if (!hasMounted) {
        sethasMounted(true)
      }
      const posX = Number(gsap.getProperty(dotContainerRef.current, 'x'))
      const landed = Math.ceil(posX / SPACER)
      onAnimationComplete(iconPaths[Math.abs(landed)].name)
    }

    const handleClick = (index: number) => {
      gsap.to([dotContainerRef.current, iconContainerRef.current], {
        duration: 0.8,
        x: snapArray[index],
        onUpdate: handleDragSlider,
        onComplete: handleThrowComplete,
        ease: 'power1',
      })
    }

    useEffect(() => {
      iconPaths.map((_, index: number) => {
        setSnapArray((snapArray) => [...snapArray, -index * SPACER])
        gsap.set(iconRefs[index], {
          transformOrigin: '50% 50%',
          scale: 0,
        })
      })

      setTimeout(() => {
        gsap.set('svg', {
          visibility: 'visible',
        })
      }, 5)

      iconPaths.map((_, index: number) => {
        const tl = gsap
          .timeline()
          .to(dotRefs[index], {
            duration: 1,
            attr: {
              r: DOT_SIZE * MULTIPLIER,
            },
            ease: 'linear',
          })
          .to(
            iconRefs[index],

            {
              duration: 1,
              alpha: 1,
              scale: 2,
              ease: 'linear',
            },
            '-=1'
          )
          .to(dotRefs[index], {
            duration: 1,
            attr: {
              r: DOT_SIZE,
            },
            ease: 'linear',
          })
          .to(
            iconRefs[index],
            {
              duration: 1,
              alpha: 0,
              scale: 0,
              ease: 'linear',
            },
            '-=1'
          )
          .time(2)

        setTls((tls) => [...tls, tl])
      })
    }, [])

    useEffect(() => {
      tls.map((tl, index: number) => mtl.add(tl, index / 2))
      Draggable.create(dotContainerRef.current, {
        type: 'x',
        bounds: {
          minX: MIN_DRAG_X,
          maxX: 0,
          minY: 0,
          maxY: 0,
        },
        onDrag: handleDragSlider,
        onDragStart: handleDragStart,
        onThrowUpdate: handleDragSlider,
        onThrowComplete: handleThrowComplete,
        inertia: true,
        minDuration: 1,
        snap: snapArray,
        overshootTolerance: 0,
        dragClickables: true,
      })
      if (!hasMounted) {
        gsap.to([dotContainerRef.current, iconContainerRef.current], 1, {
          x: snapArray[6],
          onUpdate: handleDragSlider,
          ease: 'elastic((1, 0.85)',
        })
      }
    }, [mtl])

    return (
      <svg
        ref={svgRef as RefObject<any>}
        className="svg"
        width={scale}
        height={scale}
        viewBox={`0,0, ${VIEWBOX_WIDTH}, 150`}
      >
        <defs>
          <filter id="goo" colorInterpolationFilters="sRGB">
            <feGaussianBlur in="SourceGraphic" stdDeviation="8" result="blur" />
            <feColorMatrix
              in="blur"
              mode="matrix"
              values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 31 -12"
              result="cm"
            />
          </filter>
        </defs>
        <g
          className="dot-group"
          transform={`matrix(1,0,0,1,${VIEWBOX_WIDTH / 2},${ICON_SIZE * 2})`}
        >
          <g
            ref={dotContainerRef as RefObject<any>}
            className="dot-container"
            filter="url(#goo)"
          >
            <rect
              className="hit-area"
              width={VIEWBOX_WIDTH}
              transform={`matrix(1,0,0,1,-${DOT_SIZE * 2},-${ICON_SIZE * 2})`}
            />
            {iconPaths.map((_, index: number) => (
              <circle
                ref={(ref) => {
                  dotRefs.push(ref as any)
                }}
                key={index}
                className="dot"
                cx={index * SPACER}
                cy={ICON_SIZE / 2}
                r={DOT_SIZE}
                fill={DOT_FILL}
                onClick={() => handleClick(index)}
              />
            ))}
          </g>
          <g
            ref={iconContainerRef as RefObject<any>}
            className="icon-container"
          >
            {iconPaths.map(
              (icon: { name: string; path: string }, index: number) => {
                const { name, path } = icon
                return (
                  <path
                    ref={(ref) => {
                      iconRefs.push(ref as any)
                    }}
                    key={index}
                    className="icon"
                    fill={ICON_FILL}
                    id={name}
                    data-index={index}
                    d={path}
                    opacity={0}
                    transform={`matrix(1,0,0,1,${
                      index * SPACER - ICON_SIZE / 2
                    },0)`}
                  />
                )
              }
            )}
          </g>
        </g>
      </svg>
    )
  }
)

Svg.displayName = 'Svg'
