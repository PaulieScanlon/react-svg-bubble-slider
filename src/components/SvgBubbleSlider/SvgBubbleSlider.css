.svg-bubble-slider {
  /* background-color: rgb(255, 193, 193); */
  display: flex;
  flex-direction: column;
  justify-content: center;
}
.svg {
  /* cursor: move; */
  /* width: 100%; */
  height: 100%;
  margin: 0 auto;
}

.icon {
  pointer-events: none;
}

.dot {
  cursor: pointer;
}
