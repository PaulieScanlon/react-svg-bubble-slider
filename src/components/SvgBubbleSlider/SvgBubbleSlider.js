import React from "react";

export const SvgBubbleSlider = () => {
  return <div>SvgBubbleSlider</div>;
};
