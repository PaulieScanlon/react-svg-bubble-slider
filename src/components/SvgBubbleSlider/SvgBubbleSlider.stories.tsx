import React from 'react'
import { SvgBubbleSlider } from '.'

import { Button } from 'theme-ui'

export default {
  title: 'Props',
  parameters: {
    component: SvgBubbleSlider,
    componentSubtitle: "Hooray it's a SvgBubbleSlider",
  },
}

export const usage = () => <SvgBubbleSlider />

export const action = () => (
  <SvgBubbleSlider>
    {({ reaction }) =>
      reaction && (
        <Button
          sx={{ textTransform: 'capitalize' }}
        >{`Submit ${reaction}`}</Button>
      )
    }
  </SvgBubbleSlider>
)

action.story = {
  parameters: {
    docs: {
      storyDescription: 'The `scale` prop can be used to set the size',
    },
  },
}

export const scale = () => <SvgBubbleSlider scale="50%" />

scale.story = {
  parameters: {
    docs: {
      storyDescription: 'The `scale` prop can be used to set the size',
    },
  },
}
