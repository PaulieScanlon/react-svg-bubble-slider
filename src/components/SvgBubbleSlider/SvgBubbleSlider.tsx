import React, { FunctionComponent } from 'react'

export interface SvgBubbleSliderProps {
  /** Some Prop */
  someProp?: string
}

export const SvgBubbleSlider: FunctionComponent<SvgBubbleSliderProps> = ({
  someProp,
}: SvgBubbleSliderProps) => {
  return (
    <div
      style={{
        backgroundColor: 'darkorchid',
        color: 'white',
        fontFamily: 'Roboto, Arial, sans-serif',
        padding: 10,
      }}
    >{`SvgBubbleSlider ${someProp ? someProp : ''}`}</div>
  )
}
