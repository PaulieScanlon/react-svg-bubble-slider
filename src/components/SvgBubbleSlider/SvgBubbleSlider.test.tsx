import { render } from '@testing-library/react'

import { usage } from './SvgBubbleSlider.stories'

describe('<SvgBubbleSlider />', () => {
  test('renders the component', () => {
    const { asFragment } = render(usage())
    expect(asFragment()).toMatchSnapshot()
  })
})
