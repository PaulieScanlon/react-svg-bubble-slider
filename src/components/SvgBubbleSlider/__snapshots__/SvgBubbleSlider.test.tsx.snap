// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SvgBubbleSlider /> renders the component 1`] = `
<DocumentFragment>
  <svg
    class="SvgBubbleSlider "
  >
    <circle
      class="dot"
      cx="0"
      cy="0"
      fill="hotpink"
      r="30"
    />
  </svg>
</DocumentFragment>
`;
